test/Microsoft.Azure.ServiceBus.UnitTests/Diagnostics/SessionDiagnosticsTests.cs

```diff
@@ -0,0 +1,332 @@
+        {
+            TimeSpan timeout = TimeSpan.FromSeconds(5);
+            this.queueClient = new QueueClient(TestUtility.NamespaceConnectionString,
+                TestConstants.SessionNonPartitionedQueueName, ReceiveMode.ReceiveAndDelete, new NoRetry())
+            {
+                OperationTimeout = timeout
+            };
+
+            this.queueClient.ServiceBusConnection.OperationTimeout = timeout;
+            this.queueClient.SessionClient.OperationTimeout = timeout;
+
+            Stopwatch sw = Stopwatch.StartNew();
+            ManualResetEvent processingDone = new ManualResetEvent(false);
+            this.listener.Enable((name, queueName, arg) => !name.Contains("AcceptMessageSession") &&
+                                                      !name.Contains("Receive") &&
+                                                      !name.Contains("Exception"));
+            var sessionId = Guid.NewGuid().ToString();
+            var message = new Message
+            {
+                MessageId = "messageId",
+                SessionId = sessionId
+            };
+            await this.queueClient.SendAsync(message);
+
+            this.queueClient.RegisterSessionHandler((session, msg, ct) =>
+                {
+                    processingDone.Set();
+                    return Task.CompletedTask;
+                },
+                exArgs => Task.CompletedTask);
+
+            processingDone.WaitOne(TimeSpan.FromSeconds(maxWaitSec));
+
+            Assert.True(events.TryDequeue(out var sendStart));
+            AssertSendStart(sendStart.eventName, sendStart.payload, sendStart.activity, null);
+
+            Assert.True(events.TryDequeue(out var sendStop));
+            AssertSendStop(sendStop.eventName, sendStop.payload, sendStop.activity, sendStart.activity);
+
+            Assert.True(events.TryDequeue(out var processStart));
+            AssertProcessSessionStart(processStart.eventName, processStart.payload, processStart.activity,
+                sendStart.activity);
+
+            int wait = 0;
+            while (wait++ < maxWaitSec && events.Count < 1)
+            {
+                await Task.Delay(TimeSpan.FromSeconds(1));
+            }
+
+            Assert.True(events.TryDequeue(out var processStop));
+            AssertProcessSessionStop(processStop.eventName, processStop.payload, processStop.activity,
+                processStart.activity);
+
+            Assert.True(events.IsEmpty);
+
+            // SessionPumpTaskAsync calls AcceptMessageSessionAsync() without cancellation token.
+            // Even after SessionPump is stopped, this Task may still wait for session during operation timeout
+            // It may interferee with other tests by acception it's sessions and throwing exceptions.
+            // So, let's wait for timeout and a bit more to make sure all created tasks are completed
+            sw.Stop();
+
+            await Task.Delay((int)(timeout - sw.Elapsed).TotalMilliseconds + 1000);
```

Comment:
> await Task.Delay((int)(timeout - sw.Elapsed).TotalMilliseconds + 1000); [](start = 11, length = 72)

Should have caught this before during the code review.
In every test we need to ensure we have a finally block which will do `await queueClient.CloseAsync()` and similarly for other clients

Reply:
Remove the `Task.Delay`. If the issue still persists, then its a client bug that we need to fix. Once `CloseAsync()` is invoked, it is expected that every thread it spawns is closed.

Reply:
I close queueClient and other resources in Dispose method - executed after every xunit test.
It's easier than clean up in every test.

If I remove this, tests become unstable. I've tried to fix it with a small timeout, so `AcceptMessageSessionAsync` returns faster, but it seems it's not always enough and I've seen tests fail because of it.

I can create an issue for that and add comment that this workaround should be removed once the issue is fixed. In the meantime, tests will be stable. Is it ok?

Reply:
created https://github.com/Azure/azure-service-bus-dotnet/issues/372